File: LeetCode/Medium/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
from typing import List

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left, right = 0, 0
        longest_len, cur_len = 0, 0
        char_set = set()

        while left < len(s) and right < len(s):
            
            if s[right] in char_set:
                char_set.remove(s[left])
                cur_len -= 1
                left += 1
                continue

            char_set.add(s[right])
            cur_len += 1
            right += 1
            longest_len = max(cur_len, longest_len)

        return longest_len

File: LeetCode/Medium/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
import java.util.HashSet;
import java.util.Set;

class Solution {
        public int lengthOfLongestSubstring(String s) {

            int left = 0;
            int right = 0;
            int longestLen = 0;
            int curLen = 0;
            Set<Character> charSet = new HashSet<>();

            while(left < s.length() && right < s.length()) {

                if(charSet.contains(s.charAt(right))) {
                    charSet.remove(s.charAt(left));
                    curLen--;
                    left++;
                    continue;
                }

                charSet.add(s.charAt(right));
                curLen++;
                right++;
                longestLen = Math.max(curLen, longestLen);
            }


            return longestLen;
        }
    }